File: csharp/1071.greatest-common-divisor-of-strings.cs
/*
 * @lc app=leetcode id=1071 lang=csharp
 *
 * [1071] Greatest Common Divisor of Strings
 */

// @lc code=start
public class Solution {
    public string GcdOfStrings(string str1, string str2) {
        if (!string.Equals(str1 + str2, str2 + str1)) {
            return "";
        }
        if (str1.Length == str2.Length) {
            return str1;
        }
        if (str1.Length > str2.Length) {
            return GcdOfStrings(str1.Substring(str2.Length), str2);
        }
        return GcdOfStrings(str1, str2.Substring(str1.Length));
    }
}
// @lc code=end

File: java/1071.greatest-common-divisor-of-strings.java
/*
 * @lc app=leetcode id=1071 lang=java
 *
 * [1071] Greatest Common Divisor of Strings
 */

// @lc code=start
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        String l, r;
        if (str1.length() > str2.length()) {
            l = str1;
            r = str2;
        } else {
            l = str2;
            r = str1;
        }
        int rl = r.length();
        for (int i = 0; i < rl; ++i) {
            if (l.charAt(i) != r.charAt(i)) {
                return "";
            }
        }
        if (l.length() == rl) {
            return l;
        }
        return gcdOfStrings(r, l.substring(rl));
    }
}
// @lc code=end

